targets/TARGET_STM/stm_spi_api.c

```diff
@@ -181,58 +181,135 @@ static void _spi_init_direct(spi_t *obj, const spi_pinmap_t *pinmap)
+#if defined (RCC_SPI123CLKSOURCE_PLL)
+        PeriphClkInit.Spi123ClockSelection = RCC_SPI123CLKSOURCE_PLL;
+#else
+        PeriphClkInit.Spi2ClockSelection = RCC_SPI2CLKSOURCE_SYSCLK;
+#endif
+        if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInit) != HAL_OK) {
+            error("HAL_RCCEx_PeriphCLKConfig\n");
+        }
+#endif
+
+        __HAL_RCC_SPI2_FORCE_RESET();
+        __HAL_RCC_SPI2_RELEASE_RESET();
         __HAL_RCC_SPI2_CLK_ENABLE();
         spiobj->spiIRQ = SPI2_IRQn;
     }
 #endif
 
 #if defined SPI3_BASE
     if (spiobj->spi == SPI_3) {
+#if defined(SPI_IP_VERSION_V2)
+        PeriphClkInit.PeriphClockSelection = RCC_PERIPHCLK_SPI3;
+#if defined (RCC_SPI123CLKSOURCE_PLL)
+        PeriphClkInit.Spi123ClockSelection = RCC_SPI123CLKSOURCE_PLL;
+#else
+        PeriphClkInit.Spi3ClockSelection = RCC_SPI3CLKSOURCE_SYSCLK;
+#endif
+        if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInit) != HAL_OK) {
+            error("HAL_RCCEx_PeriphCLKConfig\n");
+        }
+#endif
+
+        __HAL_RCC_SPI3_FORCE_RESET();
+        __HAL_RCC_SPI3_RELEASE_RESET();
         __HAL_RCC_SPI3_CLK_ENABLE();
         spiobj->spiIRQ = SPI3_IRQn;
     }
 #endif
 
 #if defined SPI4_BASE
     if (spiobj->spi == SPI_4) {
+#if defined(SPI_IP_VERSION_V2)
+        PeriphClkInit.PeriphClockSelection = RCC_PERIPHCLK_SPI4;
+        PeriphClkInit.Spi45ClockSelection = RCC_SPI45CLKSOURCE_PCLK1;
+        if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInit) != HAL_OK) {
+            error("HAL_RCCEx_PeriphCLKConfig\n");
+        }
+#endif
+
+        __HAL_RCC_SPI4_FORCE_RESET();
+        __HAL_RCC_SPI4_RELEASE_RESET();
         __HAL_RCC_SPI4_CLK_ENABLE();
         spiobj->spiIRQ = SPI4_IRQn;
     }
 #endif
 
 #if defined SPI5_BASE
     if (spiobj->spi == SPI_5) {
+#if defined(SPI_IP_VERSION_V2)
+        PeriphClkInit.PeriphClockSelection = RCC_PERIPHCLK_SPI5;
+        PeriphClkInit.Spi45ClockSelection = RCC_SPI45CLKSOURCE_PCLK1;
+        if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInit) != HAL_OK) {
+            error("HAL_RCCEx_PeriphCLKConfig\n");
+        }
+#endif
+
+        __HAL_RCC_SPI5_FORCE_RESET();
+        __HAL_RCC_SPI5_RELEASE_RESET();
         __HAL_RCC_SPI5_CLK_ENABLE();
         spiobj->spiIRQ = SPI5_IRQn;
     }
 #endif
 
 #if defined SPI6_BASE
     if (spiobj->spi == SPI_6) {
+#if defined(SPI_IP_VERSION_V2)
+        PeriphClkInit.PeriphClockSelection = RCC_PERIPHCLK_SPI6;
+        PeriphClkInit.Spi6ClockSelection = RCC_SPI6CLKSOURCE_PCLK4;
+        if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInit) != HAL_OK) {
+            error("HAL_RCCEx_PeriphCLKConfig\n");
+        }
+#endif
+
+        __HAL_RCC_SPI6_FORCE_RESET();
+        __HAL_RCC_SPI6_RELEASE_RESET();
         __HAL_RCC_SPI6_CLK_ENABLE();
         spiobj->spiIRQ = SPI6_IRQn;
     }
 #endif
 
     // Configure the SPI pins
     pin_function(pinmap->mosi_pin, pinmap->mosi_function);
-    pin_mode(pinmap->mosi_pin, PullNone);
+    pin_mode(pinmap->mosi_pin, PullDown);  // Pull Down is set for output line
```

Comment:
this should be in a separate commit ?

Reply:
@LMESTM 
Rebase in multi commit on top of #14981 done.

@vznncv please check in your env

Thx

Reply:
> @vznncv please check in your env

I have checked my environment (BMX160 3-wire SPI with STM32F103/STM32F411CE/STM32H743VI). I confirm that 3-wire SPI  works after these changes.